### PR TITLE
refator: FileType Autocmd and path configuration

### PR DESCRIPTION
**Pull request by tris203**
Problem:
* Autocmd: Using the BufRead autocmd means that the buffer must be focused in order to attach the LSP. As part of the integration with rzls for Razor support a virtual buffer is opened headlessly. This means that the LSP does not attach until the user BufEnters
* Lsp Cmd: Currenly it was not possible to add extra flags to the LSP start when using a mason installation.

Solution:
* Autocmd: Change to FileType Autocmd
* Lsp Cmd: Add wrapper to allow extension of Mason supplied path with either a table of a string of arguments

### CODE REVIEW

**Comment by seblj** [comment]
The autocmd makes sense. If I remember correctly, I actually had FileType autocmd previously, but changed to BufEnter. I don't really remember why. I need to take a look back in the commit history a bit first to see if there was a good reason as to why🤔 It does make sense to use FileType autocmd though, so I think that is good.

I am not sure what you mean that it is not possible to add extra flags to the server when using mason installation? Do you have a specific example in mind?

Can you not just do something like?

```lua
{
    exe = { "~/.local/share/nvim/mason/bin/roslyn", "--someflag", "--extraflag" },
}
```

**Comment by tris203** [comment]
You could but it required the user to specify the mason path, this lets you pass just the overrides
It was just sugar to make that easier

**Comment by tris203** [comment]
Also this was where you changed it
https://github.com/seblj/roslyn.nvim/pull/34

it appears it was to do with switching Solutions backwards and forwards?

**Comment by seblj** [comment]
> You could but it required the user to specify the mason path, this lets you pass just the overrides
It was just sugar to make that easier

Ahaa okay, I think I prefer not adding this sugar...

> Also this was where you changed it
https://github.com/seblj/roslyn.nvim/pull/34
>
> it appears it was to do with switching Solutions backwards and forwards?

Yeah okay I remember something like that🤔 We should definitely use FileType autocmd I think. I think it should not be a big problem with the `CSTarget` autocmd. That is actually deprecated in favor of `Roslyn target`. It does the same thing, but as I see you are aware of based on the diff, I have implemented `Roslyn restart`. So in case that the usercmd is not created, one could always just restart, and then I believe it should be available.

I could potentially also just always make the command available and then just notify the user that there is no target to change between in those cases. That is maybe the easiest way. One of the reasons I went with wanting to "hide" the usercmd if not possible to change targets was to avoid polluting the commandspace. But now it is as a subcommand to `Roslyn`

**Comment by tris203** [comment]
Thanks for understanding.

We are working on the razor support, but roslyn and rzls go hand in hand.
We require some flags passed to roslyn and to  add a couple of custom handlers for various interlinking processes.

I suspect we might find more stuff that we can work together on and upstream!

**Comment by seblj** [comment]
Sounds good! One idea though for the args. I think it could be possible to just add an `args` option to the config, where the default values are the loglevel and logdirectory. Then we could merge the user provided config into the default ones, and override if they want another loglevel for example. This should make it a bit easier and less of a special case for mason

**Comment by tris203** [comment]
> Sounds good! One idea though for the args. I think it could be possible to just add an `args` option to the config, where the default values are the loglevel and logdirectory. Then we could merge the user provided config into the default ones, and override if they want another loglevel for example. This should make it a bit easier and less of a special case for mason

I can do that if you prefer, i was just trying to avoid options overload and any sort of breaking change

**Comment by seblj** [comment]
Yeah I actually think it should be possible to add that without it being a breaking change🤔

File: lua/roslyn/init.lua [hunk 1]
```diff
@@ -275,6 +277,7 @@ function M.setup(config)
     local default_config = {
         filewatching = true,
         exe = nil,
+        extra_args = { "--logLevel=Information", "--extensionLogDirectory=" .. vim.fs.dirname(vim.lsp.get_log_path()) },
```
Comment (seblj) [inline lua/roslyn/init.lua:280]: I think it could just be called `args`

File: lua/roslyn/init.lua [hunk 2]
```diff
@@ -164,20 +164,22 @@ local function get_cmd(exe)
...
             ),
-        }, default_lsp_args)
+        }, extra_args)
     end
 end
 
 ---@class InternalRoslynNvimConfig
 ---@field filewatching boolean
 ---@field exe? string|string[]
+---@field extra_args? string[]
```
Comment (seblj) [inline lua/roslyn/init.lua:174]: This should be required for the internal type

File: lua/roslyn/init.lua [hunk 3]
```diff
@@ -144,18 +144,18 @@ local function lsp_start(cmd, bufnr, root_dir, roslyn_config, on_init)
...
 
 ---@param exe string|string[]
+---@param extra_args string[]
 ---@return string[]
-local function get_cmd(exe)
-    local default_lsp_args =
-        { "--logLevel=Information", "--extensionLogDirectory=" .. vim.fs.dirname(vim.lsp.get_log_path()) }
+local function get_cmd(exe, extra_args)
     local mason_installation = get_mason_installation()
+    local mason_exists = vim.fn.executable(mason_installation) == 1
```
Comment (seblj) [inline lua/roslyn/init.lua:151]: I don't have any reservations against extracting this to a variable, but we should still use uv.fs_stat I think

**Review by seblj — CHANGES_REQUESTED** [review]
Yes I was thinking pretty much exactly like this! Just a couple of small comments

File: lua/roslyn/init.lua [hunk 4]
```diff
@@ -87,6 +87,8 @@ local function lsp_start(cmd, bufnr, root_dir, roslyn_config, on_init)
             for _, buf in ipairs(buffers) do
                 vim.lsp.util._refresh("textDocument/diagnostic", { bufnr = buf })
             end
+
+            vim.api.nvim_exec_autocmds("User", { pattern = "RoslynInitialized", modeline = false })
```
Comment (tris203) [inline lua/roslyn/init.lua:91]: I have also added this, so we can hook it in rzls. I hope this is okay

Comment (seblj) [reply lua/roslyn/init.lua:91]: Yes of course!

**Review by seblj — APPROVED** [review]
Looks good to me! I'll just do a quick check of how this affects the usercmd before I merge. Will try to do this during the weekend

**Comment by seblj** [comment]
Yeah I think this is good! I will push a commit after merging this that just always have the user command available to be sure🤷‍♂️

**Pull request merged**